File: modules/common-logic/src/main/java/logic/Resolvable.java
package logic;

import java.util.Optional;
import logic.context.DataContext;

public interface Resolvable {

    static Resolvable just(String value) {
        return StaticResolvable.of(ResolvedValue.of(value));
    }

    static Resolvable just(int value) {
        return StaticResolvable.of(ResolvedValue.of(value));
    }

    static Resolvable just(boolean value) {
        return StaticResolvable.of(ResolvedValue.of(value));
    }

    static Resolvable just(ResolvedValue value) {
        return StaticResolvable.of(value);
    }

    static Resolvable empty() {
        return StaticResolvable.of(null);
    }

    default ResolvedValue resolve() {
        return resolve(DataContext.EMPTY);
    }

    ResolvedValue resolve(DataContext context);
}

File: pathfinder-frontend/src/logic/Resolvable.ts
import {DataContext} from "./DataContext";
import ResolvedValue from "./ResolvedValue";

interface Resolvable {

  resolve(context?: DataContext): ResolvedValue|undefined;

  asFormula(): string;
}

abstract class Resolvable {

  static just(value: string|number|boolean|ResolvedValue): Resolvable {
    if (value instanceof ResolvedValue) {
      return new Resolvable.StaticResolvable(value);
    }
    return new Resolvable.StaticResolvable(ResolvedValue.of(value));
  }

  private static StaticResolvable = class extends Resolvable {
    constructor(private readonly resolved: ResolvedValue) {
      super();
    }

    resolve(context?: DataContext): ResolvedValue | undefined {
      return this.resolved;
    }

    asFormula(): string {
      return this.resolved.asText();
    }
  }

  static readonly True = new Resolvable.StaticResolvable(ResolvedValue.True);
  static readonly False = new Resolvable.StaticResolvable(ResolvedValue.False);
}

export default Resolvable;